# Step 1

In [2]:
import pandas as pd
import numpy as np
import os

In [3]:
path=r'/Users/coley/Desktop/Data Analyst School/Achievement 4/Instacart Basket Analysis'

In [4]:
path

'/Users/coley/Desktop/Data Analyst School/Achievement 4/Instacart Basket Analysis'

In [5]:
df_ords_prods_merge=pd.read_pickle(os.path.join(path,'02 Data','Prepared Data','orders_products_merged_from_4_8.pkl'))

# Step 2

In [6]:
df_ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# Step 3

the above data and the data from the subset in the exercise differ since the data above contains all order numbers so some of the results are higher or lower when compared to the data from the subset

example: department 19 has a higher mean (17.177343) in the above data compared to the subset data (16.899756) and department 21 has a lower mean (22.902379) in the above data compared to the subset date (25.535479) - this shows that the subset data was not as accurate with producing the mean since the majority of order numbers was not included in the subset

# Step 4

In [7]:
df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [8]:
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [9]:
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [10]:
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [11]:
df_ords_prods_merge.head()

Unnamed: 0  product_id                product_name  aisle_id  \
0           0           1  Chocolate Sandwich Cookies        61   
1           0           1  Chocolate Sandwich Cookies        61   
2           0           1  Chocolate Sandwich Cookies        61   
3           0           1  Chocolate Sandwich Cookies        61   
4           0           1  Chocolate Sandwich Cookies        61   

   department_id  prices  order_id  user_id eval_set  order_number  ...  \
0             19     5.8   3139998      138    prior            28  ...   
1             19     5.8   1977647      138    prior            30  ...   
2             19     5.8    389851      709    prior             2  ...   
3             19     5.8    652770      764    prior             1  ...   
4             19     5.8   1813452      764    prior             3  ...   

   days_since_prior_order  add_to_cart_order  reordered  _merge  \
0                     3.0                  5          0    both   
1                    20.0                  1          1    both   
2                     6.0                 20          0    both   
3                     NaN                 10          0    both   
4                     9.0                 11          1    both   

     price_range_loc     busiest day  busiest days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regular days           Most Orders   
1  Mid-range product  Regularly busy  Regular days        Average Orders   
2  Mid-range product     Busiest day  Busiest days        Average Orders   
3  Mid-range product  Regularly busy  Slowest days           Most Orders   
4  Mid-range product      Least busy  Slowest days        Average Orders   

  max_order      loyalty_flag  
0        32  Regular customer  
1        32  Regular customer  
2         5      New customer  
3         3      New customer  
4         3      New customer  

[5 rows x 22 columns]

# Step 5

In [12]:
df_ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

# Step 6

In [13]:
df_ords_prods_merge['spending'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [14]:
df_ords_prods_merge.loc[df_ords_prods_merge['spending'] < 10, 'spending_flag'] = 'Low Spender'

In [15]:
df_ords_prods_merge.loc[df_ords_prods_merge['spending'] >= 10, 'spending_flag'] = 'High Spender'

In [16]:
df_ords_prods_merge['spending_flag'].value_counts(dropna=False)

Low Spender     31770746
High Spender      634113
Name: spending_flag, dtype: int64

In [17]:
df_ords_prods_merge[['user_id', 'spending_flag', 'spending']].head(10)

user_id spending_flag  spending
0      138   Low Spender  6.935811
1      138   Low Spender  6.935811
2      709   Low Spender  7.930208
3      764   Low Spender  4.972414
4      764   Low Spender  4.972414
5      777   Low Spender  6.935398
6      825   Low Spender  5.957576
7      910   Low Spender  6.680000
8     1052   Low Spender  7.162500
9     1052   Low Spender  7.162500

# Step 7

In [18]:
df_ords_prods_merge['frequency'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [19]:
df_ords_prods_merge.loc[df_ords_prods_merge['frequency'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [20]:
df_ords_prods_merge.loc[(df_ords_prods_merge['frequency'] > 10) & (df_ords_prods_merge['frequency'] <= 20), 'order_frequency_flag'] = 'Regular customer'

In [21]:
df_ords_prods_merge.loc[df_ords_prods_merge['frequency'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [22]:
df_ords_prods_merge['order_frequency_flag'].value_counts(dropna=False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_frequency_flag, dtype: int64

In [23]:
df_ords_prods_merge[['user_id', 'order_frequency_flag', 'frequency']].head(100)

user_id   order_frequency_flag  frequency
0       138      Frequent customer        8.0
1       138      Frequent customer        8.0
2       709      Frequent customer        8.0
3       764      Frequent customer        9.0
4       764      Frequent customer        9.0
..      ...                    ...        ...
95    10831  Non-frequent customer       23.0
96    11119       Regular customer       13.0
97    11186      Frequent customer        6.0
98    11243      Frequent customer        6.0
99    11448       Regular customer       14.0

[100 rows x 3 columns]

In [25]:
df_ords_prods_merge.to_pickle(os.path.join(path,'02 Data','Prepared Data','orders_products_merged_from_task_4_8.pkl'))